sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Sample.java

```diff
@@ -209,29 +202,67 @@ public void populateDisplayData(DisplayData.Builder builder) {
   }
 
   /**
-   * A {@link DoFn} that returns up to limit elements from the side input PCollection.
+   * A {@link DoFn} that outputs up to limit elements.
    */
-  private static class SampleAnyDoFn<T> extends DoFn<Void, T> {
-    long limit;
-    final PCollectionView<Iterable<T>> iterableView;
+  private static class SampleAnyDoFn<T> extends DoFn<T, T> {
```

Comment:
Is this DoFn really needed? I'm wondering if your implementation of the combiner is sufficient for performance.

Reply:
Probably not, but that might change the semantics a bit. Without this, we'll combine each window into into `n` or fewer elements. But I guess that's more natural?

Reply:
OTOH the iterable view approeach won't even work with unbounded collection so guess this new semantic is better.

Reply:
Not sure why you say that: views are also per-window, so I think it shouldn't matter whether the collection is bounded or unbounded. (though, of course, it'll be behaving weirdly in case of multiple trigger firings - see also https://issues.apache.org/jira/browse/BEAM-2305, maybe similar issues apply here too)

Reply:
With the current implementation, I got `java.lang.IllegalArgumentException: Attempted to get side input window for GlobalWindow from non-global WindowFn` with the following snippet. Am I doing something wrong?

```java
public class Test {
  public static void main(String[] args) {
    PipelineOptions options = PipelineOptionsFactory.create();
    Pipeline p = Pipeline.create(options);

    p.apply(Create.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12))
        .apply(ParDo.of(new DoFn<Integer, Integer>() {
          @ProcessElement
          public void processElement(ProcessContext c) {
            Integer x = c.element();
            c.outputWithTimestamp(x, new Instant(x * 10000));
          }
        }))
        .apply(Window.into(FixedWindows.of(Duration.standardMinutes(1))))
        .apply(Sample.any(3));
    p.run();
  }
}
```

Reply:
This seems like (another) bug in Sample.any. It should use Combine.globally().withoutDefaults() rather than simply Combine.globally().

Reply:
There's no combine in the current implementation though. It looks like this:

```java
    @Override
    public PCollection<T> expand(PCollection<T> in) {
      PCollectionView<Iterable<T>> iterableView = in.apply(View.<T>asIterable());
      return in.getPipeline()
          .apply(Create.of((Void) null).withCoder(VoidCoder.of()))
          .apply(ParDo.of(new SampleAnyDoFn<>(limit, iterableView)).withSideInputs(iterableView))
          .setCoder(in.getCoder());
    }
```

Reply:
Oh sorry, I looked in the wrong place in the code. Yeah, this implementation is wrong: the main element (null in this case) is in the global window, so of course it's not allowed to access a side input that is windowed non-globally, because it's ambiguous which window should be accessed. It plays even more poorly with triggering. See https://beam.apache.org/documentation/programming-guide/#side-inputs-and-windowing . All the more reason to fix this!

Reply:
OK, so then I'd like again to raise the suggestion to remove this DoFn :) Combining each window into n or fewer elements is the right semantics.

Reply:
Done